mantidproject/mantid
# Bug fix to prevent unhandled exception when opening fit property browser on bin plots

**Description of work.**

This fixes the issue of: "Clicking "Fit" in plot on workspace with 1 bin raises an unhandled exception" noted by @dtasev.

This error actually occurs when trying to open the fit property browser on any plot that includes a bin plot. In a bin plot the spectrum number in fact refers to the bin numbers to be plot. This causes an error when opening the fit property browser as the `getIndexFromSpectrumNumber`  function is called which tries to return a spectrum index using this number.

An input to the `getIndexFromSpectrumNumber` function is the return value of the `_get_allowed_spectra` function. This function has been changed so that any spectrum numbers related to bin plots (which are actually bin numbers) will not be returned. The `getIndexFromSpectrumNumber` will therefore not erroneously try and find the spectra index numbers of these bin numbers.

For a standalone bin plot the following error message is therefore returned to the console when clicking the fit button (due to zero spectra being returned): `"Cannot open fitting tool: No valid workspaces to fit to".`

For a overplotted spectra/bin plot, the fit button works as normal, only considering the spectra plot.


**To test:**

1. Import data (HRP39182 from training course data used).
2. Right click on workspace -> show data
3. Create a bin plot by selecting a desired number of columns, right click the header -> plot bin (values only)
4. Click fit, confirm no unhandled error.
5. Overplot desired spectra (probably no application, but its possible so...)
6. Click fit, confirm no unhandled error and fitting behaves as expected.

Fixes #31363 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
